compiler/passes/src/common/symbol_table/mod.rs

```diff
@@ -37,12 +37,16 @@ pub struct SymbolTable {
     /// The parent scope if it exists.
     /// For example, the parent scope of a then-block is the scope containing the associated ConditionalStatement.
     pub(crate) parent: Option<Box<SymbolTable>>,
-    /// Functions represents the name of each function mapped to the AST's function definition.
+    /// Functions represents the name of each local function and optional parent program name mapped to the AST's function definition.
     /// This field is populated at a first pass.
     pub functions: IndexMap<Symbol, FunctionSymbol>,
+    /// Maps parent program name and external function name to the AST's function definition.
+    pub external_functions: IndexMap<(Symbol, Symbol), FunctionSymbol>,
     /// Maps struct names to struct definitions.
     /// This field is populated at a first pass.
     pub structs: IndexMap<Symbol, Struct>,
+    /// Maps external struct names to struct definitions.
+    pub external_structs: IndexMap<(Symbol, Symbol), Struct>,
```

Comment:
Can we unify the local and external objects?
If we introduce a canonicalization pass that adds the full path to all functions, records, and structs.
In other words, local records are also specified as `<local_name>.aleo/<record_name>`.
Then you could use one mapping.
Would this be hard to do in this PR or should it be follow on work?

Reply:
Two possible simpler ways: (1) Originally it was `IndexMap<(Option<Symbol>, Symbol)>`. This simplified the code, but I ran into an issue because of json serialization of `Option<...>`. I could write a custom serialization trait for the symbol table though. (2) I can just change the Struct type and CallExpression AST type to just carry the program name instead of only carrying the program name when its external

Reply:
I suspect option 2 would be the cleanest to implement.
It also has the added benefit of making the AST more explicit.